numpy/_core/src/multiarray/stringdtype/dtype.c

```diff
@@ -0,0 +1,935 @@
+/* The implementation of the StringDType class */
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#include "structmember.h"
+
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+
+#include "numpy/arrayobject.h"
+#include "numpy/ndarraytypes.h"
+#include "numpy/npy_math.h"
+
+#include "static_string.h"
+#include "dtypemeta.h"
+#include "dtype.h"
+#include "casts.h"
+#include "gil_utils.h"
+#include "conversion_utils.h"
+#include "npy_import.h"
+
+/*
+ * Internal helper to create new instances
+ */
+PyObject *
+new_stringdtype_instance(PyObject *na_object, int coerce)
+{
+    PyObject *new =
+            PyArrayDescr_Type.tp_new((PyTypeObject *)&PyArray_StringDType, NULL, NULL);
+
+    if (new == NULL) {
+        return NULL;
+    }
+
+    npy_string_allocator *allocator = NULL;
+
+    char *default_string_buf = NULL;
+    char *na_name_buf = NULL;
+
+    allocator = NpyString_new_allocator(PyMem_RawMalloc, PyMem_RawFree,
+                                        PyMem_RawRealloc);
+    if (allocator == NULL) {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Failed to create string allocator");
+        goto fail;
+    }
+
+    npy_static_string default_string = {0, NULL};
+    npy_static_string na_name = {0, NULL};
+
+    Py_XINCREF(na_object);
+    ((PyArray_StringDTypeObject *)new)->na_object = na_object;
+    int has_null = na_object != NULL;
+    int has_nan_na = 0;
+    int has_string_na = 0;
+    if (has_null) {
+        // first check for a string
+        if (PyUnicode_Check(na_object)) {
+            has_string_na = 1;
+            Py_ssize_t size = 0;
+            const char *buf = PyUnicode_AsUTF8AndSize(na_object, &size);
+            if (buf == NULL) {
+                goto fail;
+            }
+            default_string.buf = PyMem_RawMalloc(size);
+            if (default_string.buf == NULL) {
+                PyErr_NoMemory();
```

Comment:
Should just be `goto fail;`, I think

Reply:
Thanks for noticing I missed that `goto`! It does still need to set an error though since `PyMem_RawMalloc` doesn't.